Fix Backspace deleting entire previous line in `<textarea>`

elomscansio opened the pull request:
This PR addresses issue **#27523**, where pressing backspace at the beginning of a line in a `<textarea>` incorrectly deletes the entire contents of the previous line instead of just merging the lines.  

---

### **Current Behavior**  
- When the cursor is at the start of the second line and backspace is pressed, the first line **disappears entirely** instead of merging with the second line.  

### **Expected Behavior (Per Standard)**  
- Pressing backspace at the start of a line should **move the cursor to the previous line**, deleting the newline character but **preserving the previous line's text**.  

---

### **Root Cause**  
- The function `adjust_vertical(-1, select)` modifies `selection_origin` and `edit_point`.  
- After the vertical adjustment, `self.edit_point.index = self.current_line_length();` was set without restoring `selection_origin`.  
- This led to an incorrect selection range, causing `replace_operation` to erase the entire previous line instead of just removing the newline.  

---

### **Fix**  
- Temporarily store `selection_origin` before adjusting vertical position.  
- Restore `selection_origin` after setting `edit_point.index`.  
- Ensure proper cursor movement and selection behavior.  
- Modify how backspace handles newline characters in multi-line text input to ensure only the **newline character** is removed when backspace is pressed at the start of a line while **preserving the previous line’s text**.  

---

### **Tasks**  
- [x] Reproduce the issue with tests 
- [x] Modify text deletion logic for backspace at line start  
- [x] Update test expectations  

---

### **Tests**  
- Added `test_backspace_in_textarea_at_beginning_of_line` to reproduce the issue and verify the fix.  

#### **How to Test?**  
Run:  
```sh
./mach test-unit -p script
```  
**Expected Outcome:** Backspace should merge lines without deleting text incorrectly.  

---

### **References**  
- Issue: #27523  
- [HTML Spec - Textarea Behavior](https://html.spec.whatwg.org/multipage/forms.html#the-textarea-element)

---

### **Checklist**  
- [x] `./mach build -d` does not report any errors  
- [x] `./mach test-tidy` does not report any errors  
- [x] These changes fix #27523  
- [x] There are tests for these changes

elomscansio commented:
@jdm, I noticed that `self.adjust_vertical(-1, select);` modifies both `selection_origin` and `edit_point`. However, after calling `adjust_vertical()`, we update `self.edit_point.index = self.current_line_length();` **without also updating `selection_origin`**.  

This leads to `self.selection_start()` returning `selection_origin`, while `self.selection_end()` is based on the updated `edit_point.index`. This discrepancy contributes to `replace_operation` **erasing the entire line** instead of just adjusting the cursor position.  

Would it make sense to **update `selection_origin` after setting `self.edit_point.index`** to maintain consistency? Or should we handle this differently to avoid unintended deletions?

simonwuelker commented:
(If you don't mind I'll mark this as a draft since you've posted a list of remaining TODOs, and I think the code is not meant to be reviewed yet. Feel free to change it back if I'm misunderstanding something.)

elomscansio committed:
test(textinput): Add test for backspace at beginning of line in textarea

Introduce a test to reproduce and verify the fix for backspacing at the
beginning of a line in a multiline textarea. This ensures that pressing
Backspace when the cursor is at the start of a line correctly removes the
newline without deleting the entire previous line’s content.

Related to: #27523

Signed-off-by: Emmanuel Elom <elomemmanuel007@gmail.com>

elomscansio committed:
fix(textinput): Preserve selection origin when adjusting vertical position

Fixes an issue where pressing Backspace at the beginning of a line in a
textarea incorrectly deleted the entire previous line's content. This happened
because `self.adjust_vertical(-1, select)` modified `selection_origin` and
`edit_point`, but `selection_origin` was not restored before performing the
horizontal adjustment. As a result, `self.selection_start()` and
`self.selection_end()` were inconsistent, leading to `replace_operation`
erasing the entire line.

Now, we temporarily store `selection_origin` before adjusting vertical
position and restore it afterward to ensure proper cursor and selection
behavior.

Fixes: #27523
Signed-off-by: Emmanuel Elom <elomemmanuel007@gmail.com>

elomscansio commented:
Hi @jdm @simonwuelker @gterzian  this PR is ready for review

elomscansio commented on components/script/textinput.rs:
```diff
@@ -671,8 +671,14 @@ impl<T: ClipboardProvider> TextInput<T> {
                     self.adjust_vertical(-1, select);
                     self.edit_point.index = self.current_line_length();
+                    // Restore the original selection origin to maintain expected behavior.
+                    self.selection_origin = selection_origin_temp;
```
First, when the selection origin is **greater** than the edit point, it means we are highlighting **from right to left** (backward selection). If I understand the question correctly, the scenario is:  
- The **cursor (edit point)** is at the start of the second line.  
- Some text is selected.  
- The user presses **Shift + Left Arrow** twice.  

In this case, since we are extending the selection **backward**, the **selection origin should remain unchanged** while the cursor (edit point) moves left. That is, **only the cursor moves**, not the selection origin. This is consistent with how both **Shift + Left Arrow** and **Backspace** interact with selection—except that Backspace **replaces the selected text with an empty string** after selection is made.  

However, if you are referring to **forward selection** (where the user initially selects text moving from left to right):  
- The **selection origin** would be at the **start of the second line**.  
- The **cursor (edit point)** would be somewhere **beyond** the start of the second line, depending on how much text is selected.  
- Pressing **Shift + Left Arrow** once moves the cursor to the **start of the second line**, without changing the selection origin.  
- Pressing **Shift + Left Arrow** a second time moves the cursor to the **end of the first line**, again without modifying the selection origin.  

Now, the issue before the fix was that **`adjust_vertical(-1, select)` modified `selection_origin`**, which is incorrect in the case of **backward selection**. This is the only scenario where we **needed to restore** the previous `selection_origin`, ensuring that it correctly remains at the **start of the second line**.  

This fix ensures that **selection behaves correctly** when navigating across lines, maintaining expected cursor and selection origin behavior.

jdm replied:
Ah, yeah, you're right. For some reason I was imagining that the selection origin would move when making horizontal adjustments, but that would be very strange.

elomscansio commented:
Can i proceed with another issue while awaiting for this to  be approved

jdm commented:
@elomscansio Yes, that's fine.

jdm commented:
That's https://github.com/servo/servo/issues/10083 but it probably needs to wait for #35830 .

jdm committed:
Merge branch 'main' into main

elomscansio commented:
> That's #10083 but it probably needs to wait for #35830 .

alright, though we can just delegate to the implementation in [htmlinputelement](https://github.com/servo/servo/blob/30a2a89d160dd40da6344a554ad285bb8d3d5e14/components/script/dom/htmlinputelement.rs#L2576), am I right
